Docs: Add docs for Table Maintenance in Flink

JeonDaehong opened the pull request:
This PR adds documentation for issue
( Docs update for Flink Table Maintenance API - Fixes #13307 )

It incorporates feedback received from the previously closed PR #13360. and #13771

Please review and let me know if any changes are needed. I will update promptly.

JeonDaehong commented:
@pvary 

I have incorporated the feedback you provided.

I would appreciate it if you could review it.

Where can I check the addition of orphan file cleanup in version 1.10.0?

pvary commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,363 @@
+### Why Locks Are Needed
+- **Concurrent Access**: Multiple Flink jobs may attempt maintenance simultaneously
+- **Data Consistency**: Ensures only one maintenance operation runs per table at a time
+- **Resource Management**: Prevents resource conflicts and scheduling issues
```
And wasted effort of compacting files which are compacted by another job

JeonDaehong replied:
<img width="640" height="72" alt="image" src="https://github.com/user-attachments/assets/bdeb69af-3852-4765-a9ad-063ea09fc047" />

I’ve applied it as shown in the image. Could it be that it didn’t get applied properly?

pvary commented on docs/docs/flink-table-maintenance.md:
````diff
@@ -0,0 +1,363 @@
+For single-job scenarios or testing:
+
+```java
+TriggerLockFactory lockFactory = TriggerLockFactory.defaultLockFactory();
````
I don't remember this code. Where is this coming from?

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,363 @@
+
+| Method | Description | Default Value | Type |
+|--------|-------------|---------------|------|
+| `maxSnapshotAge(Duration)` | Maximum age of snapshots to retain | None | Duration |
```
Default value is 5 days.

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,363 @@
+| Method | Description | Default Value | Type |
+|--------|-------------|---------------|------|
+| `maxSnapshotAge(Duration)` | Maximum age of snapshots to retain | None | Duration |
+| `retainLast(int)` | Minimum number of snapshots to retain | None | int |
```
Default value is 1.

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,363 @@
+| `retainLast(int)` | Minimum number of snapshots to retain | None | int |
+| `deleteBatchSize(int)` | Number of files to delete in each batch | 1000 | int |
+| `planningWorkerPoolSize(int)` | Number of worker threads for planning snapshot expiration | Shared worker pool | int |
+| `cleanExpiredMetadata(boolean)` | Remove expired metadata files when expiring snapshots | true | boolean |
```
I think in this place default value is false.

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,363 @@
+| `targetFileSizeBytes(long)` | Target size for rewritten files | Table property or 512MB | long |
+| `minFileSizeBytes(long)` | Minimum size of files eligible for compaction | 0 | long |
+| `maxFileSizeBytes(long)` | Maximum size of files eligible for compaction | Long.MAX_VALUE | long |
+| `minInputFiles(int)` | Minimum number of files to trigger rewrite | Integer.MAX_VALUE | int |
```
I think the default value is 5.

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,363 @@
+| `maxFileGroupSizeBytes(long)` | Maximum total size of a file group | 107374182400 (100GB) | long |
+| `maxFilesToRewrite(int)` | Maximum number of files to rewrite per task | Integer.MAX_VALUE | int |
+| `partialProgressEnabled(boolean)` | Enable partial progress commits | false | boolean |
+| `partialProgressMaxCommits(int)` | Maximum commits for partial progress | 10 | int |
```
`Maximum commits for partial progress, if partial progress is enabled`

JeonDaehong commented:
@pvary 

I’ve updated the document according to your feedback.

The Default Lock section was incorrectly added, so I’ve removed it.

Thank you for your review.

JeonDaehong commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,363 @@
+
+## Lock Management
+
+The `TriggerLockFactory` is essential for coordinating maintenance tasks across multiple Flink jobs or instances. It prevents concurrent maintenance operations on the same table, which could lead to conflicts or data corruption.
```
Sorry

I had already incorporated that part, but it seems to have been rolled back. I’ve applied it again!

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,361 @@
+) WITH (
+  'connector' = 'iceberg',
+  'catalog-name' = 'my_catalog',
+  'namespace' = 'db',
+  'table' = 'tbl',
```
Key should be `catalog-database` and `catalog-table`

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
````diff
@@ -0,0 +1,361 @@
+    jdbcProps                                   // JDBC connection properties
+);
+```
+
````
Can we add some specific configuration item explanations, such as `INIT_LOCK_TABLES_PROPERTY`, since this configuration item is mentioned in the SQL part below but does not appear in the Java part? https://github.com/apache/iceberg/blob/0478ff7dc8e4c048e0a1101df11656a00b9f81cd/flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/JdbcLockFactory.java#L48

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,361 @@
+- Adjust `deleteBatchSize` based on storage performance
+- Enable `partialProgressEnabled` for large rewrite operations
+- Set reasonable `maxRewriteBytes` limits
+
```
May be we can add `Setting an appropriate maxFileGroupSizeBytes can break down large FileGroups into smaller ones, thereby increasing the speed of parallel processing.`

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -0,0 +1,361 @@
+| `partialProgressEnabled(boolean)` | Enable partial progress commits | false | boolean |
+| `partialProgressMaxCommits(int)` | Maximum commits allowed for partial progress when partialProgressEnabled is true | 10 | int |
+| `maxRewriteBytes(long)` | Maximum bytes to rewrite per execution | Long.MAX_VALUE | long |
+
```
Miss a method `filter`

JeonDaehong commented:
@pvary @Guosmilesmile

Hello, I’ve incorporated the feedback you mentioned above.

This is my first time writing a Docs from start to finish, so while I’m enjoying it, I also feel like I might be causing some trouble, haha...

That said, please continue to provide feedback. I really appreciate it! :D

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -171,16 +172,17 @@ env.execute("Table Maintenance Job");
 | `partialProgressEnabled(boolean)` | Enable partial progress commits | false | boolean |
 | `partialProgressMaxCommits(int)` | Maximum commits allowed for partial progress when partialProgressEnabled is true | 10 | int |
 | `maxRewriteBytes(long)` | Maximum bytes to rewrite per execution | Long.MAX_VALUE | long |
+| `filter(Expression)` | Filter expression for selecting files to rewrite | null | Expression |
```
Default value is `Expressions.alwaysTrue()`

Guosmilesmile commented on docs/docs/flink-table-maintenance.md:
```diff
@@ -171,16 +172,17 @@ env.execute("Table Maintenance Job");
 | `rewriteAll(boolean)` | Rewrite all data files regardless of thresholds | false | boolean |
 | `maxFileGroupSizeBytes(long)` | Maximum total size of a file group | 107374182400 (100GB) | long |
-| `maxFilesToRewrite(int)` | Maximum number of files to rewrite per task | Integer.MAX_VALUE | int |
+| `maxFilesToRewrite(int)` | About the description we can add: If this option is not specified, all eligible files will be rewritten | null | int |
```
We don't need `About the description we can add: ` in the doc.

Guosmilesmile reviewed (commented):
I have left some mirror comments. I think we can reuse the existing page https://iceberg.apache.org/docs/nightly/flink-actions/ instead of adding a new file. We can rename flink-actions to flink-maintenance, which would eliminate the need for a new page.

JeonDaehong commented:
> I have left some mirror comments. I think we can reuse the existing page https://iceberg.apache.org/docs/nightly/flink-actions/ instead of adding a new file. We can rename flink-actions to flink-maintenance, which would eliminate the need for a new page.

Thank you for the feedback. Once I revise the parts you mentioned and add them to the Flink action docs, would that be all that’s needed?

Guosmilesmile commented:
> > I have left some mirror comments. I think we can reuse the existing page https://iceberg.apache.org/docs/nightly/flink-actions/ instead of adding a new file. We can rename flink-actions to flink-maintenance, which would eliminate the need for a new page.
> 
> Thank you for the feedback. Once I revise the parts you mentioned and add them to the Flink action docs, would that be all that’s needed?

I don’t see any other changes needed for now. Once you finish the revisions, I will review it again. Thanks for your hard work!

JeonDaehong commented:
@Guosmilesmile 

How should we handle the 'Rewrite Files' action in flink-action?

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -18,7 +18,7 @@ title: "Flink Actions"
  -->
 
-## Rewrite files action
+# Rewrite files action
```
```suggestion
## Flink Table Maintenance BatchMode
### Rewrite files action
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+
+# Flink Table Maintenance 
+
+## Overview
```
```suggestion
### Overview
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+
+The `TableMaintenance` API in **Apache Iceberg** empowers **Flink jobs** to execute maintenance tasks **natively**, either embedded within existing streaming pipelines or deployed as standalone Flink jobs. This eliminates dependencies on external systems, thereby **streamlining architecture**, **reducing operational costs**, and **enhancing automation capabilities**.
+
+## Supported Features (Flink)
```
```suggestion
### Supported Features (Flink)
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+
+## Supported Features (Flink)
+
+### ExpireSnapshots
```
```suggestion
#### ExpireSnapshots
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+    .deleteBatchSize(1000))
+```
+
+### RewriteDataFiles
````
```suggestion
#### RewriteDataFiles
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+    .partialProgressMaxCommits(5))
+```
+
+## Lock Management
````
```suggestion
### Lock Management
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+
+The `TriggerLockFactory` is essential for coordinating maintenance tasks. It prevents concurrent maintenance operations on the same table, which could lead to conflicts or data corruption. This locking mechanism is necessary even for a single job, as multiple instances of the same task could otherwise conflict.
+
+### Why Locks Are Needed
```
```suggestion
#### Why Locks Are Needed
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+- **Resource Management**: Prevents resource conflicts and scheduling issues
+- **Avoid Duplicate Work**: Even when only a single compaction job is scheduled, multiple instances could attempt the same operation, leading to redundant work and wasted resources.
+
+### Supported Lock Types
```
```suggestion
#### Supported Lock Types
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+
+### Supported Lock Types
+
+#### JDBC Lock Factory
```
```suggestion
##### JDBC Lock Factory
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+);
+```
+
+#### ZooKeeper Lock Factory
````
```suggestion
##### ZooKeeper Lock Factory
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+);
+```
+
+## Quick Start
````
```suggestion
### Quick Start
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+env.execute("Table Maintenance Job");
+```
+
+## Configuration Options
````
```suggestion
### Configuration Options
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+
+## Configuration Options
+
+### TableMaintenance Builder
```
```suggestion
#### TableMaintenance Builder
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+| `parallelism(int)` | Default parallelism for maintenance tasks | System default |
+| `maxReadBack(int)` | Max snapshots to check during initialization | 100 |
+
+### Maintenance Task Common Options
```
```suggestion
#### Maintenance Task Common Options
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+| `scheduleOnEqDeleteRecordCount(long)` | Trigger after N equality delete records | No automatic scheduling | long |
+| `scheduleOnInterval(Duration)` | Trigger after time interval | No automatic scheduling | Duration |
+
+### ExpireSnapshots Configuration
```
```suggestion
#### ExpireSnapshots Configuration
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+| `planningWorkerPoolSize(int)` | Number of worker threads for planning snapshot expiration | Shared worker pool | int |
+| `cleanExpiredMetadata(boolean)` | Remove expired metadata files when expiring snapshots | false | boolean |
+
+### RewriteDataFiles Configuration
```
```suggestion
#### RewriteDataFiles Configuration
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+| `maxRewriteBytes(long)` | Maximum bytes to rewrite per execution | Long.MAX_VALUE | long |
+| `filter(Expression)` | Filter expression for selecting files to rewrite | Expressions.alwaysTrue() | Expression |
+
+## Complete Example
```
```suggestion
### Complete Example
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+}
+```
+
+## IcebergSink with Post-Commit Integration
````
```suggestion
### IcebergSink with Post-Commit Integration
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+    .append();
+```
+
+### SQL Examples
````
```suggestion
#### SQL Examples
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+);
+```
+
+## Lock Configuration (SQL)
````
```suggestion
### Lock Configuration (SQL)
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+| `flink-maintenance.lock.zookeeper.max-retries` | Max retries | `3` |
+| `flink-maintenance.lock.zookeeper.base-sleep-ms` | Base sleep between retries (ms) | `3000` |
+
+## Best Practices
```
```suggestion
### Best Practices
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+
+## Best Practices
+
+### Resource Management
```
```suggestion
#### Resource Management
```

Guosmilesmile commented on docs/docs/flink-actions.md:
```diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+- Set appropriate parallelism based on cluster resources
+- Enable checkpointing for fault tolerance
+
+### Scheduling Strategy
```
```suggestion
#### Scheduling Strategy
```

Guosmilesmile commented on docs/docs/flink-actions.md:
````diff
@@ -33,3 +33,359 @@ RewriteDataFilesActionResult result = Actions.forTable(table)
+.rateLimit(Duration.ofMinutes(10))     // Reduce frequency of task execution
+```
+
+### Slow rewrite operations
````
```suggestion
#### Slow rewrite operations
```

Guosmilesmile reviewed (commented):
1. `flink-configuration.md`should not be within the scope of the current changes. Could it have been accidentally deleted?
2. May be we should also change the `title: "Flink Actions"` in flink-actions.md to `title: "Flink TableMaintenance"`, and rename `flink-actions.md` to `flink-maintenance.md` , then change the mkdocs.yml ?

JeonDaehong commented:
It looks like I mistakenly deleted Maintenance when I should have just removed it. I’ll restore it.

I’ll make the corrections you mentioned today and push the update again.

Thank you for the feedback!

Guosmilesmile commented:
@JeonDaehong Sorry, I thought you had already submitted the changes when I saw your reply. Just take your time.

JeonDaehong commented:
> @JeonDaehong Sorry, I thought you had already submitted the changes when I saw your reply. Just take your time.

It seems the Korean → English translation didn’t come out quite right. Haha..

pvary commented on docs/docs/flink-maintenance.md:
````diff
@@ -0,0 +1,393 @@
+
+```java
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+TableLoader tableLoader = TableLoader.fromHadoopTable("path/to/table");
````
`fromHadoopTable` -> Can we change this to `hive` or something else?
We would like to remove the references of the HadoopTable from the docs as it is not encouraged in production environments

JeonDaehong commented:
@pvary 

Can we change Hadoop to Hive like this?

pvary commented on docs/docs/flink-maintenance.md:
```diff
@@ -0,0 +1,401 @@
+import org.apache.iceberg.flink.actions.Actions;
+
+TableLoader tableLoader = TableLoader.fromCatalog(
+    CatalogLoader.hive("my_catalog", configuration),
```
`CatalogLoader.hive("my_catalog", configuration, properties),`

pvary commented:
Merged to main.
Thanks for the doc @JeonDaehong and @Guosmilesmile for the review!

JeonDaehong commented:
@pvary 

Has this Docs been visible on the Docs website since the 1.9.3 release?
